lib/pbench/server/filetree.py

```diff
@@ -298,20 +327,84 @@ def unpack(self, incoming: Path, results: Path):
         ensured by calling this indirectly through the Controller class unpack
         method.
 
-        TODO: This is a prototype for testing, and doesn't actually unpack
-        the tarball as we're going to be relying on the current unpack pipeline
-        command for some time.
-
         Args:
             incoming: Controller's directory in the INCOMING tree
             results: Controller's directory in the RESULTS tree
+
+        Raises:
+            TarballUnpackError if an exception/failure occurs while unpacking
+            the tarball.
         """
-        self.logger.warning("Tarball unpack is not yet implemented")
-        unpacked = incoming / self.name
-        unpacked.mkdir(parents=True)  # Just create an empty directory for now
-        self.unpacked = unpacked
+        unpacked = incoming / f"{self.name}.unpack"
+        unpacked.mkdir(parents=True)
+
+        try:
+            process = subprocess.run(
+                [
+                    "tar",
+                    "-x",
+                    "--no-same-owner",
+                    "--delay-directory-restore",
+                    "--force-local",
+                    "--file=" + str(self.tarball_path),
+                ],
+                cwd=unpacked,
+                stdin=subprocess.DEVNULL,
+                capture_output=True,
+                text=True,
+            )
+        except Exception as exc:
+            raise TarballUnpackError(self.tarball_path, str(exc)) from exc
+        else:
+            if process.returncode != 0:
+                raise TarballUnpackError(
```

Comment:
Also here, and several other places.

One possibility would be to just have an outer `try`/`except` block *after* setting `unpacked`: it would remove the `unpacked` directory tree on any exception if the directory exists, and re-raise the original exception.

Although it might be better to do this as a `try`/`finally` since we *always* want to remove `unpacked` even on success. (And on an error, this will also take whatever we might have unpacked underneath `unpacked`.)